### PR TITLE
Avoid creating unnecessary tables for Resource and DomainResource resource types #713

### PR DESCRIPTION
**Pull request by prb112**
- Checks for DOMAINRESOURCE and RESOURCE tables view
- Removes the artifacts
- Bonus: added fix for null condition on a code in DefaultMemberMatchStrategy
- Updated the automation to remove the `Resource` (including Build Scripts for various CI Tests)

### CODE REVIEW

File: fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/app/Main.java [hunk 1]
```diff
@@ -453,39 +458,38 @@ protected void updateFhirSchema() {
...
+
                 // V0014 IS_DELETED and LAST_UPDATED added to whole-system LOGICAL_RESOURCES
                 applyDataMigrationForV0014();
-        
-                // Log warning messages that unused tables will be removed in a future release.
-                // TODO: This will no longer be needed after the tables are removed (https://github.com/IBM/FHIR/issues/713).
-                logWarningMessagesForDeprecatedTables();
-                
+
+                // V0021 removes Abstract Type tables which are unused.s
```
Comment (tbieste) [inline fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/app/Main.java:482]: "unused.s" typo

File: fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/MigrateV0021AbstractTypeRemoval.java [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+    // Total number of Tables that have data.
+    private int count = 0;
+
+    /**
+     * Public constructor
+     *
+     * @param adapter
+     * @param adminSchemaName
+     * @param schemaName
+     * @param force - indicating if their is data in these tables, it'll still remove the table.
```
Comment (tbieste) [inline fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/MigrateV0021AbstractTypeRemoval.java:72]: "their is" typo

File: fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/MigrateV0021AbstractTypeRemoval.java [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+        }
+        cleanupHistory(translator, c);
+    }
+
+    /**
+     * Controls the throwing of Exceptions.
+     */
+    private void checkShouldThrowException() {
+        if (count != 0 && !force) {
+            throw new IllegalArgumentException("The Data Tables for Resources or DomainResource contains data. use --force-unused-table-removal to force removal");
```
Comment (tbieste) [inline fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/MigrateV0021AbstractTypeRemoval.java:105]: "use" -> "Use"

**Review by tbieste — COMMENTED** [review]
I like the "force" option to allow the admin to choose to proceed easily.

File: fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/UnusedTableRemovalNeedsV0021Migration.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+     * @return
+     */
+    public boolean checkDerby(IDatabaseTranslator translator, Connection c) {
+        // Grab the list of tables for the configured schema from the Derby sys catalog
+        final String sql = ""
+                + "SELECT 1 FROM sys.systables AS tables "
+                + "  JOIN sys.sysschemas AS schemas "
+                + "    ON (tables.schemaid = schemas.schemaid) "
+                + " WHERE schemas.schemaname = ?"
+                + "     AND schemas.tablename in (" + addParameterMarkers(TABLE_COUNT) + ")";
```
Comment (prb112) [inline fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/control/UnusedTableRemovalNeedsV0021Migration.java:111]:
```suggestion
                + "     AND tables.tablename in (" + addParameterMarkers(TABLE_COUNT) + ")";
```

File: build/derby-bootstrap.sh [hunk 5]
```diff
@@ -26,13 +26,13 @@ java -jar ${WORKSPACE}/fhir-persistence-schema/target/fhir-persistence-schema-*-
   --update-schema
 java -jar ${WORKSPACE}/fhir-persistence-schema/target/fhir-persistence-schema-*-cli.jar \
   --db-type derby --prop db.database=${DB_LOC}/profile --prop db.create=Y \
-  --prop resourceTypes=Patient,Group,Practitioner,PractitionerRole,Person,RelatedPerson,Organization,Location,Observation,MedicationAdministration,StructureDefinition,ElementDefinition,CodeSystem,ValueSet,Resource \
```
Comment (lmsurpre) [inline build/derby-bootstrap.sh:29]: good catch

**Review by lmsurpre — APPROVED** [review]
LGTM

**Pull request merged**